File: defmc/src/defmc_android.cpp
#include <dmsdk/sdk.h>
#include "defmc_private.h"

#if defined(DM_PLATFORM_ANDROID)

static JNIEnv* Attach()
{
	JNIEnv* env;
	JavaVM* vm = dmGraphics::GetNativeAndroidJavaVM();
	vm->AttachCurrentThread(&env, NULL);
	return env;
}

static bool Detach(JNIEnv* env)
{
	bool exception = (bool) env->ExceptionCheck();
	env->ExceptionClear();
	JavaVM* vm = dmGraphics::GetNativeAndroidJavaVM();
	vm->DetachCurrentThread();
	return !exception;
}

namespace {
	struct AttachScope
	{
		JNIEnv* m_Env;
		AttachScope() : m_Env(Attach())
		{
		}
		~AttachScope()
		{
			Detach(m_Env);
		}
	};
}

static jclass GetClass(JNIEnv* env, const char* classname)
{
	jclass activity_class = env->FindClass("android/app/NativeActivity");
	jmethodID get_class_loader = env->GetMethodID(activity_class,"getClassLoader", "()Ljava/lang/ClassLoader;");
	jobject cls = env->CallObjectMethod(dmGraphics::GetNativeAndroidActivity(), get_class_loader);
	jclass class_loader = env->FindClass("java/lang/ClassLoader");
	jmethodID find_class = env->GetMethodID(class_loader, "loadClass", "(Ljava/lang/String;)Ljava/lang/Class;");

	jstring str_class_name = env->NewStringUTF(classname);
	jclass outcls = (jclass)env->CallObjectMethod(cls, find_class, str_class_name);
	env->DeleteLocalRef(str_class_name);
	return outcls;
}

int DefMcPlatform_Init()
{
	dmLogInfo("ANDROID DefMcPlatform_Init ...");
	dmLogInfo("DONE!");

	return 0;
}

int DefMcPlatform_Start(uint32_t sample_rate, uint32_t sample_delay, float lowpass_alpha)
{
	dmLogInfo("ANDROID DefMcPlatform_Start ... rate: %u, delay: %u", sample_rate, sample_delay);

	// prepare JNI
	AttachScope attachscope;
	JNIEnv* env = attachscope.m_Env;
	jclass cls = GetClass(env, "com.defold.android.defmc.DefMcExtension");

	// get buffersize
	jmethodID get_buf_size = env->GetStaticMethodID(cls, "GetMinBufferSize", "(I)I");
	uint32_t buf_len = (uint32_t)env->CallStaticIntMethod(cls, get_buf_size, (int)sample_rate);
	dmLogInfo("buf_len: %u", buf_len);

	// start recording
	jmethodID start = env->GetStaticMethodID(cls, "StartRecorder", "(IIF)V");
	env->CallStaticVoidMethod(cls, start, (int)sample_rate, (int)sample_delay, lowpass_alpha);

	dmLogInfo("ANDROID DefMcPlatform_Start DONE");
	return 0;
}


int DefMcPlatform_Stop()
{
	dmLogInfo("ANDROID DefMcPlatform_Stop ...");
	// prepare JNI
	AttachScope attachscope;
	JNIEnv* env = attachscope.m_Env;
	jclass cls = GetClass(env, "com.defold.android.defmc.DefMcExtension");

	// call method
	dmLogInfo("GetStaticMethodID ...");
	jmethodID start = env->GetStaticMethodID(cls, "StopRecorder", "(Landroid/app/Activity;)V");
	dmLogInfo("DONE!");
	dmLogInfo("CallStaticVoidMethod ...");
	env->CallStaticVoidMethod(cls, start, dmGraphics::GetNativeAndroidActivity());
	dmLogInfo("DONE!");

	dmLogInfo("ANDROID DefMcPlatform_Stop DONE");
	return 0;
}

int DefMcPlatform_SampleAmplitude(float &amp)
{
	// prepare JNI
	AttachScope attachscope;
	JNIEnv* env = attachscope.m_Env;
	jclass cls = GetClass(env, "com.defold.android.defmc.DefMcExtension");

	// get lowpass filtered amplitude
	jmethodID start = env->GetStaticMethodID(cls, "SampleAmplitude", "(Landroid/app/Activity;)D");
	amp = (float)env->CallStaticDoubleMethod(cls, start, dmGraphics::GetNativeAndroidActivity());

	return 0;
}

#endif

File: defmc/src/defmc_android.java
package com.defold.android.defmc;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;
import android.app.Activity;
import android.content.Context;
import java.lang.Runnable;

class DefMcExtension {
	private final static String TAG = "defmc";

	private static int sampleRate = 8000;
	private static double lastLevel = 11;
	private static int sampleDelay = 75;
	private static AudioRecord audio;
	private static Thread thread;
	private static int bufferSize = 0;
	private static short buffer[];
	private static short buffer_filtered[];
	private static Boolean stopping = false;
	private static float alpha = 0.5f;

	public static void InitRecorder() { }

	public static int GetMinBufferSize(int sample_rate) {
		try {
			bufferSize = AudioRecord.getMinBufferSize(
				sample_rate, 
				AudioFormat.CHANNEL_IN_MONO,
				AudioFormat.ENCODING_PCM_16BIT);
		} catch (Exception e) {
			android.util.Log.e(TAG, "Exception", e);
		}

		return bufferSize;
	}

	public static void StartRecorder(int sample_rate, int sample_delay, float lp_alpha) {
		android.util.Log.v(TAG, "################### StartRecorder, sample_rate: "+sample_rate+", sample_delay: "+sample_delay);

		if (thread != null) {
			return;
		}

		sampleRate = sample_rate;
		sampleDelay = sample_delay;
		alpha = lp_alpha;
		stopping = false;

		buffer = new short[bufferSize];
		buffer_filtered = new short[bufferSize];
		for(int i=0; i < bufferSize; ++i) {
			buffer[i] = 0;
			buffer_filtered[i] = 0;
		}

		android.util.Log.v(TAG, "Allocating new AudioRecord...");
		audio = new AudioRecord(MediaRecorder.AudioSource.MIC, sampleRate,
				AudioFormat.CHANNEL_IN_MONO,
				AudioFormat.ENCODING_PCM_16BIT, bufferSize);
		android.util.Log.v(TAG, "DONE, starting recording...");

		thread = new Thread(new Runnable() {
			public void run() {
				audio.startRecording();
				android.util.Log.v(TAG, "DONE");
				while(thread != null && !thread.isInterrupted()) {
					if (stopping) {
						try {
							if (audio != null) {
								audio.stop();
								audio.release();
								audio = null;
							}
						} catch (Exception e) {e.printStackTrace();}
						thread.interrupt();
						thread = null;
						break;
					}

					try {Thread.sleep(sampleDelay);} catch(InterruptedException ie) {ie.printStackTrace();break;}
					try {
						int buffer_read_result = 1;
						if (audio != null) {
							buffer_read_result = audio.read(buffer, 0, bufferSize);
							double sum_level = 0;
							for (int i = 1; i < buffer_read_result; i++) {
								buffer_filtered[i] *= (short) (1.0 - alpha);
								buffer_filtered[i] += (short)(alpha * (float)(buffer[i] - buffer_filtered[i]));
								sum_level += buffer_filtered[i];
							}
							lastLevel = Math.abs((sum_level / buffer_read_result));
						}

					} catch (Exception e) {
						e.printStackTrace();
					}
				}
			}
		});
		thread.start();
	}

	public static void GetAudioBuffer(short[] nativeBuf) {
		android.util.Log.v(TAG, "##################### GetAudioBuffer");
	}

	public static void StopRecorder(final Activity activity) {
		stopping = true;
	}

	public static double SampleAmplitude(final Activity activity) {
		return lastLevel;
	}
}